include/irr/core/math/morton.h

```diff
@@ -0,0 +1,58 @@
+#ifndef __IRR_MORTON_H_INCLUDED__
+#define __IRR_MORTON_H_INCLUDED__
+
+#include <cstdint>
+#include "IrrCompileConfig.h"
+#include "irr/macros.h"
+#include "irr/static_if.h"
+
+namespace irr {
+namespace core
+{
+
+namespace impl
+{
+    template <typename T>
+    constexpr T morton2d_decode_mask(uint32_t _n)
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR uint64_t mask[5]
+        {
+            0x5555555555555555ull,
+            0x3333333333333333ull,
+            0x0F0F0F0F0F0F0F0Full,
+            0x00FF00FF00FF00FFull,
+            0x0000FFFF0000FFFFull
+        };
+        return static_cast<T>(mask[_n]);
+    }
+
+    template <typename T>
+    inline T morton2d_decode(T x)
+    {
+        x = x & morton2d_decode_mask<T>(0);
+        x = (x | (x >> 1)) & morton2d_decode_mask<T>(1);
+        x = (x | (x >> 2)) & morton2d_decode_mask<T>(2);
```

Comment:
maybe make the template
`template<typename T, size_t bitDepth=sizeof(T)*8u>`

so the "range" is independent of type (useful for having fast mortons for small quadtrees)

Reply:
still relevant

Reply:
it is, thats why `if constexpr`s are there

Reply:
you're not getting me, you're assuming that if I do a 2d morton code with a return type `uint32_t` I want a 16bit X,Y resolution

but I could want a 32bit value, while only wanting a 4bit X,Y resolution (morton code for values up to 255) - that stuff is useful for computing small neighbourhoods, etc.

This is why the bit-depth of the morton code should not be inferred from the size of the return type.

Reply:
ah, ok, now i get what u mean